src/mlpack/methods/ann/layer/linear.hpp

```diff
@@ -143,6 +143,25 @@ class LinearType : public Layer<MatType>
   template<typename Archive>
   void serialize(Archive& ar, const uint32_t /* version */);
 
+  using Layer<MatType>::As;
+
+  template<typename TargetMatType>
+  Layer<TargetMatType>* As() const
```

Comment:
Was an extra template parameter to the copy/move constructors not working?  I really suspect that is the better way to go here.